api/src/services/purged-docs.js

```diff
@@ -0,0 +1,161 @@
+  const purgedIds = [];
+  if (!result || !result.results) {
+    return purgedIds;
+  }
+
+  result.results.forEach(change => {
+    if (!change.deleted) {
+      purgedIds.push(purgingUtils.extractId(change.id));
+    }
+  });
+  return purgedIds;
+};
+
+const getPurgedIds = (roles, docIds) => {
+  if (!docIds || !docIds.length || !roles || !roles.length) {
+    return Promise.resolve([]);
+  }
+
+  const cacheKey = getCacheKey(roles, docIds);
+  const cached = cache.get(cacheKey);
+  if (cached) {
+    cache.ttl(cacheKey);
+    return Promise.resolve(cached);
+  }
+
+  const purgeDb = getPurgeDb(roles);
+  const ids = docIds.map(purgingUtils.getPurgedId);
+
+  // requesting _changes instead of _all_docs because it's roughly twice faster
+  return purgeDb
+    .changes({ doc_ids: ids, batch_size: ids.length + 1, seq_interval: ids.length })
+    .then(result => {
+      const purgeIds = getPurgedIdsFromChanges(result);
+      cache.set(cacheKey, purgeIds);
+      return purgeIds;
+    });
+};
+
+const getPurgedIdsSince = (roles, docIds, { checkPointerId = '', limit = 100 } = {}) => {
+  if (!docIds || !docIds.length || !roles || !roles.length) {
+    return Promise.resolve([]);
+  }
+
+  const purgeDb = getPurgeDb(roles);
+  const ids = docIds.map(purgingUtils.getPurgedId);
+
+  return getCheckPointer(purgeDb, checkPointerId)
+    .then(checkPointer => {
+      const opts = {
+        doc_ids: ids,
+        batch_size: ids.length + 1,
+        limit: limit,
+        since: checkPointer.last_seq,
+        seq_interval: ids.length
+      };
+
+      return purgeDb.changes(opts);
+    })
+    .then(result => {
+      const purgedDocIds = getPurgedIdsFromChanges(result);
+      return {
+        purgedDocIds,
+        lastSeq: result.last_seq
+      };
+    });
+};
+
+const getCheckPointer = (db, checkPointerId) => db
+  .get(`_local/${checkPointerId}`)
+  .catch(() => ({
+    _id: `_local/${checkPointerId}`,
+    last_seq: 0
+  }));
+
+const writeCheckPointer = (roles, checkPointerId, seq = 0) => {
+  const purgeDb = getPurgeDb(roles);
+
+  return Promise
+    .all([
+      getCheckPointer(purgeDb, checkPointerId),
+      purgeDb.info()
+    ])
+    .then(([ checkPointer, info ]) => {
+      checkPointer.last_seq = seq === 'now' ? info.update_seq : seq;
+      purgeDb.put(checkPointer);
+    });
+};
+
+const info = (roles) => {
+  const purgeDb = getPurgeDb(roles);
+  return purgeDb.info();
+};
+
+const listen = (seq = 'now') => {
+  db.sentinel
+    .changes({ live: true, since: seq })
+    .on('change', change => {
+      seq = change.seq;
+      if (change.id.startsWith('purgelog:') && change.changes[0].rev.startsWith('1-')) {
+        clearCache();
+      }
+    })
+    .on('error', err => {
+      logger.error('Error watching sentinel changes, restarting: %o', err);
+      listen(seq);
+    });
+};
+
+let inited = false;
+const init = () => {
+  if (inited) {
+    return;
+  }
+  listen();
+  inited = true;
+};
```

Comment:
Instead of an `init` function can we start the listener immediately when the module is `require`d? Again it makes testing a little harder because you have to mock before you require but it makes the code much cleaner because it'll only be invoked once and we never have to check again.

Reply:
Done. I just added an `if` not to start the listener when running tests.